packages/cli/src/lib/cache/SuccessCache.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs-extra';
+import { resolve as resolvePath } from 'node:path';
+
+const DEFAULT_CACHE_BASE_PATH = 'node_modules/.cache/backstage-cli';
+
+const CACHE_MAX_AGE_MS = 7 * 24 * 3600_000;
+
+export class SuccessCache {
+  readonly #path: string;
+
+  constructor(name: string, basePath?: string) {
+    this.#path = resolvePath(basePath ?? DEFAULT_CACHE_BASE_PATH, name);
+  }
+
+  async read(): Promise<Set<string>> {
+    const state = await fs.stat(this.#path).catch(error => {
+      if (error.code === 'ENOENT') {
+        return undefined;
+      }
+      throw error;
+    });
+    if (!state || !state.isDirectory()) {
+      return new Set();
+    }
+    const items = await fs.readdir(this.#path);
+
+    const returned = new Set<string>();
+    const removed = new Set<string>();
+
+    const now = Date.now();
+
+    for (const item of items) {
+      const split = item.split('_');
+      if (split.length !== 2) {
+        removed.add(item);
+        continue;
+      }
+      const createdAt = parseInt(split[0], 10);
+      if (Number.isNaN(createdAt) || now - createdAt > CACHE_MAX_AGE_MS) {
+        removed.add(item);
+      } else {
+        returned.add(split[1]);
+      }
+    }
+
+    for (const item of removed) {
+      await fs.unlink(resolvePath(this.#path, item));
+    }
+
+    return returned;
+  }
+
+  async write(newEntries: Iterable<string>): Promise<void> {
+    const now = Date.now();
+
+    await fs.ensureDir(this.#path);
+
+    const existingItems = await fs.readdir(this.#path);
+
+    const empty = Buffer.alloc(0);
+    for (const key of newEntries) {
+      // Remove any existing items with the key we're about to add
+      const trimmedItems = existingItems.filter(item =>
+        item.endsWith(`_${key}`),
+      );
+      for (const trimmedItem of trimmedItems) {
+        await fs.unlink(resolvePath(this.#path, trimmedItem));
+      }
+
+      await fs.writeFile(resolvePath(this.#path, `${now}_${key}`), empty);
```

Comment:
you're REALLY avoiding making like a yaml state file or so :)

Reply:
Come to think of it, i guess this actually is racy isn't it (delete and then create), if you have multiple concurrent builds going on. It's extremely unlikely of course that you'd do that AND have hash matches. :) So I'm not really worried. Otherwise you could do like, write first, then delete everything that matches with an older timestamp than that of the new entry. Or something. But whatever.

Reply:
yeah seems very unlikely in the first place, and you also wouldn't be running multiple builds of this in the first place. It explicitly prevents mixing with the shard flag for example: https://github.com/backstage/backstage/blob/f98ca6919574a9c877b805ee6a708973ff9ade19/packages/cli/src/commands/repo/test.ts#L290-L294

Reply:
Also yes, trying to keep it very simple. I did actually put the timestamp in the file content to start but realized that's gonna a be a bit of a performance hit to every time have to read through all the files to clean the cache